diff -Nuarp file-5.22-ori/src/Makefile.am file-5.22/src/Makefile.am
--- file-5.22-ori/src/Makefile.am	2013-02-22 15:03:35.000000000 +0100
+++ file-5.22/src/Makefile.am	2015-03-06 22:54:48.834569451 +0100
@@ -12,6 +12,7 @@ libmagic_la_SOURCES = magic.c apprentice
 	funcs.c file.h readelf.h tar.h apptype.c \
 	file_opts.h elfclass.h mygetopt.h cdf.c cdf_time.c readcdf.c cdf.h
 libmagic_la_LDFLAGS = -no-undefined -version-info 1:0:0
+libmagic_la_LDFLAGS += $(XC_LINKTOOL_CFLAGS)
 if MINGW
 MINGWLIBS = -lgnurx -lshlwapi
 else
