diff --git a/src/base/ftobjs.c b/src/base/ftobjs.c
index 1b042614d..80ce6a995 100644
--- a/src/base/ftobjs.c
+++ b/src/base/ftobjs.c
@@ -820,6 +820,9 @@
     library = driver->root.library;
     hinter  = library->auto_hinter;
 
+    /* NOTE: Kindle Hack. Always force autohint and grey smooth antialias */
+    load_flags |= FT_LOAD_FORCE_AUTOHINT | FT_LOAD_TARGET_NORMAL;
+
     /* resolve load flags dependencies */
 
     if ( load_flags & FT_LOAD_NO_RECURSE )
