diff -Nuarp glib-2.42.2-ori/configure.ac glib-2.42.2/configure.ac
--- glib-2.42.2-ori/configure.ac	2015-02-26 04:08:08.000000000 +0100
+++ glib-2.42.2/configure.ac	2015-03-06 21:05:09.611821054 +0100
@@ -3465,6 +3465,7 @@ AC_ARG_ENABLE(Bsymbolic,
 if test "x${enable_Bsymbolic}" = "xyes"; then
   GLIB_LINK_FLAGS=-Wl,-Bsymbolic-functions
 fi
+GLIB_LINK_FLAGS+=" ${XC_LINKTOOL_CFLAGS}"
 
 AC_SUBST(GLIB_LINK_FLAGS)
 
