diff --git a/Makefile.defs b/Makefile.defs
index 11be6068..5f2c9382 100644
--- a/Makefile.defs
+++ b/Makefile.defs
@@ -246,10 +246,10 @@ endif
 # set cross-compiler/host AR LD RANLIB
 ifdef CHOST
 	STRIP:=$(CHOST)-strip
-	AR:=$(CHOST)-ar
+	AR:=$(CHOST)-gcc-ar
 	LD:=$(CHOST)-ld
 	RC:=$(CHOST)-windres
-	RANLIB:=$(CHOST)-ranlib
+	RANLIB:=$(CHOST)-gcc-ranlib
 else
 	STRIP:=strip
 	AR:=ar
@@ -279,13 +279,13 @@ endif
 
 # Base CFLAGS, without arch. We'll need it for luajit, because its
 # Makefiles do some tricky stuff to differentiate HOST/TARGET
-BASE_CFLAGS:=-O2 -ffast-math -pipe -fomit-frame-pointer
+#BASE_CFLAGS:=-O2 -ffast-math -pipe -fomit-frame-pointer
 # For ricers.
 #BASE_CFLAGS:=-O3 -ffast-math -pipe -fomit-frame-pointer -frename-registers -fweb
 # Used to enforce aggressive optimizations in some components
 VECTO_CFLAGS:=-O3 -ffast-math -pipe -fomit-frame-pointer -frename-registers -fweb -ftree-vectorize -funroll-loops
 # And for an aggressive LTO build
-#BASE_CFLAGS:=$(VECTO_CFLAGS) -flto=8 -fuse-linker-plugin
+BASE_CFLAGS:=$(VECTO_CFLAGS) -flto=8 -fuse-linker-plugin
 # Use this for debugging:
 ifdef KODEBUG
 	BASE_CFLAGS:=-Og -g -pipe -fno-omit-frame-pointer
