diff --git a/configure.ac b/configure.ac
index 2cb006af..d993dab9 100644
--- a/configure.ac
+++ b/configure.ac
@@ -337,6 +337,7 @@ if test x$intel_compiler = xyes; then
   fi
   AC_MSG_RESULT([$have_static_libcxa])
 fi
+LDFLAGS_NOSTARTFILES+=" ${XC_LINKTOOL_CFLAGS}"
 
 if test x$qcc_compiler = xyes; then
     LDFLAGS_NOSTARTFILES="-XCClinker -Wc,-nostartfiles"
diff --git a/src/Makefile.am b/src/Makefile.am
index 8d72ef0b..fd87d44b 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -50,6 +50,7 @@ libunwind_ptrace_la_SOURCES =						  \
 	ptrace/_UPT_put_unwind_info.c ptrace/_UPT_get_proc_name.c	  \
 	ptrace/_UPT_reg_offset.c ptrace/_UPT_resume.c
 noinst_HEADERS += ptrace/_UPT_internal.h
+libunwind_ptrace_la_LDFLAGS = $(XC_LINKTOOL_CFLAGS)
 
 ### libunwind-coredump:
 libunwind_coredump_la_SOURCES = \
