diff -Nuarp openssl-1.0.2-ori/crypto/armcap.c openssl-1.0.2/crypto/armcap.c
--- openssl-1.0.2-ori/crypto/armcap.c	2015-01-22 15:58:32.000000000 +0100
+++ openssl-1.0.2/crypto/armcap.c	2015-02-17 20:24:04.905251752 +0100
@@ -137,6 +137,8 @@ void OPENSSL_cpuid_setup(void)
     } else if (sigsetjmp(ill_jmp, 1) == 0) {
         _armv7_neon_probe();
         OPENSSL_armcap_P |= ARMV7_NEON;
+        // Kindle Hack: We're on ARMv7, so nerf this to avoid spamming dmesg w/ undefined instructions...
+        /*
         if (sigsetjmp(ill_jmp, 1) == 0) {
             _armv8_pmull_probe();
             OPENSSL_armcap_P |= ARMV8_PMULL | ARMV8_AES;
@@ -152,11 +154,16 @@ void OPENSSL_cpuid_setup(void)
             _armv8_sha256_probe();
             OPENSSL_armcap_P |= ARMV8_SHA256;
         }
+        */
     }
+    // Kindle Hack: Apparently, the kernel doesn't enable this part of the PMU for userland on the PW2's A9 (at least on FW 5.4.2).
+    // Thus this throws an (handled) SIGILL. Nerf for now, to avoid spamming dmesg with 'undefined instruction' logs.
+    /*
     if (sigsetjmp(ill_jmp, 1) == 0) {
         _armv7_tick();
         OPENSSL_armcap_P |= ARMV7_TICK;
     }
+    */
 
     sigaction(SIGILL, &ill_oact, NULL);
     sigprocmask(SIG_SETMASK, &oset, NULL);
