diff -Nuarp readline-6.3-ori/shlib/Makefile.in readline-6.3/shlib/Makefile.in
--- readline-6.3-ori/shlib/Makefile.in	2012-01-12 22:54:33.000000000 +0100
+++ readline-6.3/shlib/Makefile.in	2015-03-06 17:05:17.419748215 +0100
@@ -113,7 +113,7 @@ CCFLAGS = $(DEFS) $(LOCAL_DEFS) $(CPPFLA
 
 .c.so:
 	${RM} $@
-	$(SHOBJ_CC) -c $(CCFLAGS) $(SHOBJ_CFLAGS) -o $*.o $< 
+	$(SHOBJ_CC) -c $(CCFLAGS) $(SHOBJ_CFLAGS) $(LDFLAGS) -o $*.o $< 
 	$(MV) $*.o $@
 
 # The name of the main library target.
@@ -174,17 +174,17 @@ unsupported:
 
 $(SHARED_READLINE):	$(SHARED_OBJ)
 	$(RM) $@
-	$(SHOBJ_LD) ${SHOBJ_LDFLAGS} ${SHLIB_XLDFLAGS} -o $@ $(SHARED_OBJ) $(SHLIB_LIBS)
+	$(SHOBJ_LD) $(CCFLAGS) ${SHOBJ_CFLAGS} ${SHOBJ_LDFLAGS} ${SHLIB_XLDFLAGS} -o $@ $(SHARED_OBJ) $(SHLIB_LIBS)
 
 $(SHARED_HISTORY):	$(SHARED_HISTOBJ) xmalloc.so xfree.so
 	$(RM) $@
-	$(SHOBJ_LD) ${SHOBJ_LDFLAGS} ${SHLIB_XLDFLAGS} -o $@ $(SHARED_HISTOBJ) xmalloc.so xfree.so $(SHLIB_LIBS)
+	$(SHOBJ_LD) $(CCFLAGS) ${SHOBJ_CFLAGS} ${SHOBJ_LDFLAGS} ${SHLIB_XLDFLAGS} -o $@ $(SHARED_HISTOBJ) xmalloc.so xfree.so $(SHLIB_LIBS)
 
 # Since tilde.c is shared between readline and bash, make sure we compile 
 # it with the right flags when it's built as part of readline
 tilde.so:        tilde.c
 	${RM} $@
-	$(SHOBJ_CC) -c $(CCFLAGS) $(SHOBJ_CFLAGS) -DREADLINE_LIBRARY -c -o tilde.o $(topdir)/tilde.c
+	$(SHOBJ_CC) -c $(CCFLAGS) $(SHOBJ_CFLAGS) $(LDFLAGS) -DREADLINE_LIBRARY -c -o tilde.o $(topdir)/tilde.c
 	$(MV) tilde.o $@
 
 installdirs: $(topdir)/support/mkdirs
