diff --git a/requests/__init__.py b/requests/__init__.py
index 626247cb..469b1bb0 100644
--- a/requests/__init__.py
+++ b/requests/__init__.py
@@ -48,27 +48,26 @@ from .exceptions import RequestsDependencyWarning
 
 def check_compatibility(urllib3_version, chardet_version):
     urllib3_version = urllib3_version.split('.')
-    assert urllib3_version != ['dev']  # Verify urllib3 isn't installed from git.
 
     # Sometimes, urllib3 only reports its version as 16.1.
     if len(urllib3_version) == 2:
         urllib3_version.append('0')
 
     # Check urllib3 for compatibility.
-    major, minor, patch = urllib3_version  # noqa: F811
-    major, minor, patch = int(major), int(minor), int(patch)
-    # urllib3 >= 1.21.1, <= 1.25
-    assert major == 1
-    assert minor >= 21
-    assert minor <= 25
+    if urllib3_version != ['dev']:
+        major, minor, patch = urllib3_version  # noqa: F811
+        major, minor, patch = int(major), int(minor), int(patch)
+        # urllib3 >= 1.21.1, <= 1.25
+        assert major == 1
+        assert minor >= 21
+        assert minor <= 25
 
     # Check chardet for compatibility.
     major, minor, patch = chardet_version.split('.')[:3]
     major, minor, patch = int(major), int(minor), int(patch)
-    # chardet >= 3.0.2, < 3.1.0
-    assert major == 3
-    assert minor < 1
-    assert patch >= 2
+    # chardet >= 3.0.x, <= 4.0.x
+    assert major >= 3
+    assert major <= 4
 
 
 def _check_cryptography(cryptography_version):
